fairscale/experimental/nn/offload.py

```diff
@@ -0,0 +1,447 @@
+            )
+
+        # The returned variables need to mirror the forward inputs
+        # TODO(anj-s): Why do we need to do this?
+        if isinstance(grad_outputs, tuple):
+            return grad_outputs[0], None, None, None
+
+        return grad_outputs, None, None, None
+
+
+class OffloadModel(nn.Module):
+    """Wrapper used offload parts of a model to the CPU.
+
+    The model is sharded such that parts of the model are copied
+    from CPU->GPU for the FW pass and back right after. The reverse
+    copy is carried out for the backward pass.
+
+    Args:
+        module (~torch.nn.Sequential): Module to be offloaded.
+
+        device (torch.device):
+            Device where the active model should reside.
+
+        offload_device (torch.device):
+            Device where the inactive model should reside.
+
+        num_slices (int):
+            Number of slices into which the model should be chunked.
+    """
+
+    def __init__(
+        self,
+        model_cpu: nn.Sequential,  # hard pre-requisite for now, easier model slicing
+        device: torch.device,
+        offload_device: torch.device = torch.device("cpu"),
+        num_slices: int = 5,
+        checkpoint_activation: bool = False,
+        num_microbatches: int = 1,
+    ):
+        super().__init__()
+        # TODO(anj-s): Add error checks for cuda and sequential model.
+
+        self.device = device
+        self.offload_device = offload_device
+
+        # Slice the model into roughly equivalent sequential shards.
+        splits = _split(model_cpu, num_slices)
+
+        # List of model shards that will be placed on/off the device.
+        self.model_slices: List[nn.Module] = []
+
+        for i, split in enumerate(splits):
+            # Add one model handling this slice
+            self.model_slices.append(
+                ModelShard(
+                    cpu_model_shard=nn.Sequential(*split), device=device, offload_device=offload_device, index=i,
+                )
+            )
+
+        # Expose a unified view of the slices
+        self.model = torch.nn.Sequential(*self.model_slices)
+
+        # intermediate activations at the slice boundaries.
+        self._activations: List[Tuple] = []
+
+        # Currently we only support microbatches with activation checkpointing.
+        if not checkpoint_activation and num_microbatches > 1:
+            raise RuntimeError("We currently only support microbatches with activation checkpointing.")
+
+        # Bool indicating if we want to checkpoint activation on the host.
+        self._checkpoint_activation = checkpoint_activation
+
+        # Number of microbatches to run per batch on the device
+        self._num_microbatches = num_microbatches
+
+    def forward(self, *inputs: Any, **_: Any) -> Any:
+        # At least one of the inputs needs to have `requires_grad` set.
+        # TODO(anj-s): Should we require users to set this or should we set it here?
+        for inp in inputs:
```

Comment:
I would add an exception here if none of the inputs could be tagged, so that it would explain a bit more what the issue is ?